src/NHibernate/Linq/DefaultQueryProvider.cs

```diff
@@ -15,13 +15,20 @@ namespace NHibernate.Linq
 {
 	public partial interface INhQueryProvider : IQueryProvider
 	{
+		//TODO: Obsolete? We don't really need ExecuteFuture, ExecuteFutureValue for unified batch
 		IFutureEnumerable<TResult> ExecuteFuture<TResult>(Expression expression);
 		IFutureValue<TResult> ExecuteFutureValue<TResult>(Expression expression);
 		void SetResultTransformerAndAdditionalCriteria(IQuery query, NhLinqExpression nhExpression, IDictionary<string, Tuple<object, IType>> parameters);
 		int ExecuteDml<T>(QueryMode queryMode, Expression expression);
 		Task<TResult> ExecuteAsync<TResult>(Expression expression, CancellationToken cancellationToken);
 	}
 
+	public interface INhQueryProviderSupportMultiBatch
+	{
+		IQuery GetPreparedQuery(Expression exrpession, out NhLinqExpression expression);
+		IMultiAnyQueryBatch GetFutureMultiBatch();
```

Comment:
I'm thinking to expose `ISessionImplementor Session` instead. And get batch from session. Any objections?

Reply:
I do not really see a benefit in exposing the whole session from the Linq provider. I would rather keep just exposing the future batch. Or even cease exposing it and add methods for adding new Linq query to it, a bit like the old interface was doing.

Reply:
The benefit is to make further modifications simpler. If session was available and old future was using it - we wouldn't need to introduce new interface. It was only internal modifications in extension methods. Is Session really internal implementation detail for provider? I guess it can be abstracted one more layer down - but it looks to me as very rare breaking change (and such modification will be breaking in many more places).

Reply:
Ok, go for it.